pyhf/pdf.py

```diff
@@ -70,68 +88,146 @@ def par_slice(self, name):
-
-        return modifier
+        self.paramsets.setdefault(name, [])
+
+        parset = modifier_cls.required_parset(len(sample['data']))
+        parset.update({'channel,sample': (channel['name'], sample['name'])})
+
+        if not(parset['is_shared']) and self.paramsets.get(name):
+            raise ValueError("Trying to add unshared-paramset but other paramsets exist with the same name.")
+
+        if parset['is_shared'] and self.paramsets.get(name) and not(self.paramsets.get(name)[0]['is_shared']):
+            raise ValueError("Trying to add shared-paramset but other paramset of same name is indicated to be unshared.")
+
+        self.paramsets.get(name).append(parset)
+
+    def combine_paramsets(self):
+        for param_name in list(self.paramsets.keys()):
+            params = self.paramsets[param_name]
+
+            def get_inits(parset, npars, param_matching):
+                if parset == unconstrained:
+                    if param_matching == 'exact':
+                        # normfactor
+                        inits = [1.0]
+                        bounds = [[0, 10]]
+                        auxdata = []
+                        factors = []
+                    else:
+                        # shapefactor
+                        inits = [1.0]
+                        bounds = [[0, 10]]
+                        auxdata = []
+                        factors = []
+                elif parset == constrained_by_normal:
+                    if param_matching == 'exact':
+                        # normsys, histosys
+                        inits = [0.0]
+                        bounds = [[-5., 5.]]
+                        auxdata = [0.]
+                        factors = []
+                    else:
+                        # staterror
+                        inits = [1.]
+                        bounds = [[1e-10, 10.]]
+                        auxdata = [1.]
+                        factors = []
+                elif parset == constrained_by_poisson:
+                    if param_matching == 'exact':
+                        raise ValueError("No idea how to handle Poisson constraint with exact n_parameters = 1")
+                    else:
+                        # shapesys
+                        inits = [1.0]
+                        bounds = [[1e-10, 10.]]
+                        auxdata = [-1.]
+                        factors = [-1.]
+                return {'inits': inits * npars,
+                        'bounds': bounds * npars,
+                        'auxdata': auxdata * npars,
+                        'factors': factors * npars}
+
+            if len(params) == 1:
+                self.paramsets[param_name] = self.paramsets[param_name][0]
+            else:
+                channel_sample = set([])
+                constraint_types = set([])
+                n_parameters = set([])
+                param_matching = set([])
+                mod_types = set([])
+                for param in params:
```

Comment:
not clear to me why those (except n_parameters) are needed (as they're not used by `add_paramsets`)